python/ray/util/ray_executor.py

```diff
@@ -0,0 +1,317 @@
+            Futures are aggregated into this list as they are returned.
+        self.__actor_pool:
+            ActorPool which is used by self.map() when max_workers > 1
+        self.__remote_fn:
+            Wrapper around the remote function to be executed by Ray.
+        self.context:
+            Context containing settings and attributes returned after
+            initialising the Ray client.
+        """
+
+        self._shutdown_ray: bool = shutdown_ray
+        self._shutdown_lock: bool = False
+        self._futures: List[Future] = []
+        self.__actor_pool: Optional[ActorPool] = None
+        self.__remote_fn: "Optional[RemoteFunction0]" = None
+        self.context: "Optional[BaseContext]" = None
+
+        # The following is necessary because `@ray.remote` is only available at runtime.
+        if max_workers is None:
+
+            @ray.remote
+            def remote_fn(fn: Callable[[], T]) -> T:
+                return fn()
+
+            self.__remote_fn = remote_fn
+        else:
+            if max_workers < 1:
+                raise ValueError(
+                    f"`max_workers={max_workers}` is given. The argument \
+                    `max_workers` must be >= 1"
+                )
+
+            @ray.remote
+            class ExecutorActor:
+                def __init__(self):
+                    pass
+
+                def actor_function(self, fn: Callable):
+                    return fn()
+
+                def ready(self):
+                    return
+
+            actors = [
+                ExecutorActor.options(  # type: ignore[attr-defined]
+                    name=f"actor-{i}"
+                ).remote()
+                for i in range(max_workers)
+            ]
+            for a in actors:
+                ray.get(a.ready.remote())
+            self.__actor_pool = ray.util.ActorPool(actors)
+
+        self.context = ray.init(ignore_reinit_error=True, **kwargs)
+
+    def submit(
+        self, fn: Callable[P, T], /, *args: P.args, **kwargs: P.kwargs
+    ) -> Future[T]:
+        """Submits a callable to be executed with the given arguments.
+
+        Schedules the callable to be executed as `fn(*args, **kwargs)` and returns
+        a Future instance representing the execution of the callable.
+
+        Returns:
+            A Future representing the given call.
+
+        Usage example:
+
+        .. code-block:: python
+
+            with RayExecutor() as ex:
+                future_0 = ex.submit(lambda x: x * x, 100)
+                future_1 = ex.submit(lambda x: x + x, 100)
+                result_0 = future_0.result()
+                result_1 = future_1.result()
+        """
+        self._check_shutdown_lock()
+        fn_curried = partial(fn, *args, **kwargs)
+        if self.__actor_pool:
+            self.__actor_pool.submit(
+                lambda a, _: a.actor_function.remote(fn_curried), None
+            )
+            oref = self.__actor_pool._index_to_future[
```

Comment:
I don't think we should access private fields of ActorPool. I think it might be easier to manage a pool of actors internally instead of using `ray.util.ActorPool` (like what multiprocessing.Pool does).

Reply:
actor pools have been removed